core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -900,18 +908,42 @@ private[spark] object MapOutputTracker extends Logging {
     } {
       objOut.close()
     }
-    val arr = out.toByteArray
+
+    val arr: Array[Byte] = {
+      val zos = new ZstdOutputStream(compressedOut)
+      Utils.tryWithSafeFinally {
+        compressedOut.write(DIRECT)
+        // `out.writeTo(zos)` will write the uncompressed data from `out` to `zos`
+        //  without copying to avoid unnecessary allocation and copy of byte[].
+        out.writeTo(zos)
+      } {
+        zos.close()
+      }
+      compressedOut.toByteArray
+    }
     if (arr.length >= minBroadcastSize) {
       // Use broadcast instead.
       // Important arr(0) is the tag == DIRECT, ignore that while deserializing !
       val bcast = broadcastManager.newBroadcast(arr, isLocal)
       // toByteArray creates copy, so we can reuse out
       out.reset()
-      out.write(BROADCAST)
-      val oos = new ObjectOutputStream(new GZIPOutputStream(out))
-      oos.writeObject(bcast)
-      oos.close()
-      val outArr = out.toByteArray
+      val oos = new ObjectOutputStream(out)
+      Utils.tryWithSafeFinally {
+        oos.writeObject(bcast)
+      } {
+        oos.close()
+      }
+      val outArr = {
+        compressedOut.reset()
+        val zos = new ZstdOutputStream(compressedOut)
```

Comment:
Hi, @dbtsai , I am back-porting this into our internal repo. Looks like this compression is unnecessary since `arr` is already compressed by zstd. Compress again with already compressed byte[] is a waste of cpu time. WDYT?

Reply:
The actually value of the data (which is already compressed) will not be in the serialized form of  `out.writeTo(zos)` as it's transient. Here, we are just serializing the reference to the actual data, and the actual data will be broadcast through `TorrentBroadcast`. See the next log, `"Broadcast mapstatuses size = " + outArr.length + ", actual size = " + arr.length` for your real data. The broadcast one is very small.

Reply:
Thanks for your clarification. It's indeed not including the compressed data.